File: src/lib/api/officeGateway.ts
import { invoke } from "@tauri-apps/api/core";

export type OfficeGatewayProviderKind =
  | "auto"
  | "deep_seek"
  | "kimi"
  | "mimo"
  | "mini_max";

export interface OfficeGatewayConfig {
  listenHost: string;
  listenPort: number;
  activeProvider: OfficeGatewayProviderKind;
  defaultMaxTokens: number;
  minCompatMaxTokens: number;
  passthroughMetadata: boolean;
  enableWebSearchTool: boolean;
  modelPrimary: string;
  modelMid: string;
  modelFast: string;
  deepseekApiKey: string;
  deepseekBaseUrl: string;
  deepseekModelPrimary: string;
  deepseekModelMid: string;
  deepseekModelFast: string;
  kimiApiKey: string;
  kimiPaygBaseUrl: string;
  kimiCodingBaseUrl: string;
  kimiCodingModel: string;
  kimiModelPrimary: string;
  kimiModelMid: string;
  kimiModelFast: string;
  mimoApiKey: string;
  mimoPaygBaseUrl: string;
  mimoTpRegion: string;
  mimoTpBaseUrlCn: string;
  mimoTpBaseUrlSgp: string;
  mimoTpBaseUrlAms: string;
  mimoModelPrimary: string;
  mimoModelMid: string;
  mimoModelFast: string;
  minimaxApiKey: string;
  minimaxRegion: string;
  minimaxBaseUrlCn: string;
  minimaxBaseUrlGlobal: string;
  minimaxModelPrimary: string;
  minimaxModelMid: string;
  minimaxModelFast: string;
}

export interface OfficeGatewayStatus {
  running: boolean;
  host: string;
  port: number;
  baseUrl: string;
  activeProvider: OfficeGatewayProviderKind;
  logFile: string;
  startedAt: string | null;
}

export interface OfficeGatewayLogEntry {
  ts: string;
  level: string;
  category: string;
  message: string;
}

export interface OfficeGatewayLogSnapshot {
  entries: OfficeGatewayLogEntry[];
  logFile: string;
}

export interface OfficeGatewayUpstreamTestResult {
  ok: boolean;
  provider: OfficeGatewayProviderKind;
  routeKind: string;
  upstreamUrl: string;
  model: string;
  status: number;
  message: string;
  bodyPreview: string;
}

export const officeGatewayApi = {
  getConfig: () =>
    invoke<OfficeGatewayConfig>("get_office_gateway_config"),
  saveConfig: (config: OfficeGatewayConfig) =>
    invoke<OfficeGatewayConfig>("save_office_gateway_config", { config }),
  start: () => invoke<OfficeGatewayStatus>("start_office_gateway"),
  stop: () => invoke<void>("stop_office_gateway"),
  restart: () => invoke<OfficeGatewayStatus>("restart_office_gateway"),
  getStatus: () => invoke<OfficeGatewayStatus>("get_office_gateway_status"),
  getLogs: () =>
    invoke<OfficeGatewayLogSnapshot>("get_office_gateway_logs"),
  testUpstream: () =>
    invoke<OfficeGatewayUpstreamTestResult>("test_office_gateway_upstream"),
  clearLogs: () => invoke<void>("clear_office_gateway_logs"),
  openLogFile: () => invoke<void>("open_office_gateway_log_file"),
};

File: src-tauri/src/commands/office_gateway.rs
use crate::office_gateway::{
    OfficeGatewayConfig, OfficeGatewayLogSnapshot, OfficeGatewayService, OfficeGatewayStatus,
    OfficeGatewayUpstreamTestResult,
};
use std::sync::Arc;
use tauri_plugin_opener::OpenerExt;

pub struct OfficeGatewayState(pub Arc<OfficeGatewayService>);

#[tauri::command]
pub async fn get_office_gateway_config(
    state: tauri::State<'_, OfficeGatewayState>,
) -> Result<OfficeGatewayConfig, String> {
    Ok(state.0.get_config().await)
}

#[tauri::command]
pub async fn save_office_gateway_config(
    state: tauri::State<'_, OfficeGatewayState>,
    config: OfficeGatewayConfig,
) -> Result<OfficeGatewayConfig, String> {
    state.0.save_config(config).await?;
    Ok(state.0.get_config().await)
}

#[tauri::command]
pub async fn start_office_gateway(
    state: tauri::State<'_, OfficeGatewayState>,
) -> Result<OfficeGatewayStatus, String> {
    state.0.start().await
}

#[tauri::command]
pub async fn stop_office_gateway(
    state: tauri::State<'_, OfficeGatewayState>,
) -> Result<(), String> {
    state.0.stop().await
}

#[tauri::command]
pub async fn restart_office_gateway(
    state: tauri::State<'_, OfficeGatewayState>,
) -> Result<OfficeGatewayStatus, String> {
    state.0.restart().await
}

#[tauri::command]
pub async fn get_office_gateway_status(
    state: tauri::State<'_, OfficeGatewayState>,
) -> Result<OfficeGatewayStatus, String> {
    Ok(state.0.status().await)
}

#[tauri::command]
pub async fn get_office_gateway_logs(
    state: tauri::State<'_, OfficeGatewayState>,
) -> Result<OfficeGatewayLogSnapshot, String> {
    Ok(state.0.logs().await)
}

#[tauri::command]
pub async fn test_office_gateway_upstream(
    state: tauri::State<'_, OfficeGatewayState>,
) -> Result<OfficeGatewayUpstreamTestResult, String> {
    state.0.test_upstream().await
}

#[tauri::command]
pub async fn clear_office_gateway_logs(
    state: tauri::State<'_, OfficeGatewayState>,
) -> Result<(), String> {
    state.0.clear_logs().await
}

#[tauri::command]
pub async fn open_office_gateway_log_file(
    app: tauri::AppHandle,
    state: tauri::State<'_, OfficeGatewayState>,
) -> Result<(), String> {
    let path = state.0.log_file_path();
    if !std::path::Path::new(&path).exists() {
        state.0.clear_logs().await?;
    }
    app.opener()
        .open_path(path, None::<&str>)
        .map_err(|e| format!("Failed to open Office Gateway log file: {e}"))
}
